File: test/writeFile1.c
#include <stdio.h>
#include "mpitoyogi.h"

int main(int argc, char **argv){

    int i, rank, size, offset, nints, N=16;
    MPI_File fhw;
    MPI_Status status;
    MPI_Init(&argc, &argv);
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    MPI_Comm_size(MPI_COMM_WORLD, &size);
    int buf[N];
    for ( i=0;i<N;i++){
        buf[i] = i ;
    }
    offset = rank*(N/size)*sizeof(int);
    MPI_File_open(MPI_COMM_WORLD, "datafile", MPI_MODE_CREATE|MPI_MODE_WRONLY,
                  MPI_INFO_NULL, &fhw);
    printf("\nRank: %d, Offset: %d\n", rank, offset);
    MPI_File_write_at(fhw, offset, buf, (N/size), MPI_INT, &status);
    MPI_File_close(&fhw);
    MPI_Finalize();
    return 0;
}

File: test/writeFile1.f90
program writeFile1

include 'mpif.h'

    integer :: i, rank, size, nints, N, ierr
    integer(kind=MPI_OFFSET_KIND) :: offset
    integer :: fhw, buf(16)
    integer :: status(MPI_STATUS_SIZE)

    N = 16

    call MPI_INIT(ierr)
    call MPI_COMM_RANK(MPI_COMM_WORLD, rank, ierr)
    call MPI_COMM_SIZE(MPI_COMM_WORLD, size, ierr)

    do i=1,N
        buf(i) = i - 1 
    enddo

    offset = rank*(N/size)*4
    call MPI_FILE_OPEN(MPI_COMM_WORLD, "fWriteFile.result", &
                           OR(MPI_MODE_CREATE, MPI_MODE_WRONLY), &
                           MPI_INFO_NULL, fhw, ierr)
    call MPI_FILE_WRITE_AT(fhw, offset, buf, (N/size), MPI_INTEGER, status, ierr)
    call MPI_FILE_CLOSE(fhw, ierr);
    call MPI_FINALIZE(ierr);
end program
